ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,335 @@
+    }
+    for (Map.Entry<Long, Long> sizeAndSegmentsNeeded : overallRequirements.entrySet()) {
+      long sizeInBytes = sizeAndSegmentsNeeded.getKey();
+      long origSegmentsNeeded = sizeAndSegmentsNeeded.getValue();
+      // Ensure that swap segments are only added to the requirements if the number of swap segments allocated to
+      // stores is lower than the minimum required swap segments.
+      long swapSegmentsNeeded = Math.max(requiredSwapSegmentsPerSize - swapSegmentsUsed.get(sizeInBytes), 0L);
+      overallRequirements.put(sizeInBytes, origSegmentsNeeded + swapSegmentsNeeded);
+    }
+    return Collections.unmodifiableMap(overallRequirements);
+  }
+
+  /**
+   * Create a directory if it does not yet exist and check that it is accessible.
+   * @param dir The directory to possibly create.
+   * @return the passed-in directory
+   * @throws IOException if the specified directory could not be created/read, or if a non-directory file with the same
+   *                     name already exists.
+   */
+  private static File prepareDirectory(File dir) throws IOException {
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+    if (!dir.isDirectory()) {
+      throw new IOException("Directory could not be created or is not a directory: " + dir.getAbsolutePath());
+    }
+    return dir;
+  }
+
+  /**
+   * @return a filename for a reserve file. This is reserve_ followed by a random UUID.
+   */
+  private static String generateFilename() {
+    return RESERVE_FILE_PREFIX + UUID.randomUUID();
+  }
+
+  private static class ReserveFileMap {
+    private final Map<Long, Queue<File>> internalMap = new HashMap<>();
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+
+    /**
+     * Add a file of the specified file size to the reserve file map.
+     * @param sizeInBytes the size of the reserve file
+     * @param reserveFile the reserve file to add.
+     */
+    void add(long sizeInBytes, File reserveFile) {
+      rwLock.writeLock().lock();
+      try {
+        internalMap.computeIfAbsent(sizeInBytes, key -> new LinkedList<>()).add(reserveFile);
+      } finally {
+        rwLock.writeLock().unlock();
+      }
+    }
+
+    /**
+     * Remove and return a file of the specified size from the reserve file map.
+     * @param sizeInBytes the size of the file to look for.
+     * @return the {@link File}, or {@code null} if no file exists in the map for the specified size.
+     */
+    File remove(long sizeInBytes) {
+      rwLock.writeLock().lock();
+      try {
+        File reserveFile = null;
+        Queue<File> reserveFilesForSize = internalMap.get(sizeInBytes);
+        if (reserveFilesForSize != null) {
+          if (reserveFilesForSize.size() != 0) {
+            reserveFile = reserveFilesForSize.remove();
+          }
+          if (reserveFilesForSize.size() == 0) {
+            internalMap.remove(sizeInBytes);
```

Comment:
I am thinking about removing this so that i can get rid of the rwlock and use concurrent data structures.